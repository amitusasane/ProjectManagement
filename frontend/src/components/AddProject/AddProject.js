import React, { useEffect, useState } from 'react';
import { Container, Row, Col, FormControl, Button, ListGroup, Alert, Form } from 'react-bootstrap';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import './AddProject.scss';

import { getAllProject, addNewProject, updateProjectById, deleteProjectById } from '../../api/Api';
import * as moment from 'moment';

const AddProject = () => {
  const initialStartDate = formatDate(new Date());
  const initialEndDate = formatDate(getAfterDate(1));

  let [projectUpdated, setProjectUpdated] = useState(false);
  let [projects, setProjects] = useState([]);

  let [editMode, setEditMode] = useState(false);
  let [projectId, setProjectId] = useState('');
  let [projectName, setProjectName] = useState('');

  let [startDate, setStartDate] = useState(initialStartDate);
  let [endDate, setEndDate] = useState(initialEndDate);
  let [dateRequired, setDateRequired] = useState(false);

  let [priority, setPriority] = useState('');
  let [statusMessage, setStatusMessage] = useState({ show: false, message: '', variant: '' });

  async function fetchAllProjects() {
    try {
      setProjects(await getAllProject());
      setProjectUpdated(false);
    } catch (err) {
      setStatusMessage({ ...statusMessage, show: true, message: err, variant: 'danger' });
    }
  }

  async function deleteProject(project) {
    try {
      const resp = await deleteProjectById(project);
      setProjectUpdated(true);
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: resp.message,
        variant: 'success'
      });
    } catch (err) {
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: err,
        variant: 'danger'
      });
    }
    resetFormState();
    formik.resetForm();
  }

  function editProject(project) {
    setEditMode(true);
    setProjectId(project._id);
    setProjectName(project.projectName);
    setDateRequired(project.dateRequired);
    setStartDate(project.startDate ? formatDate(project.startDate) : initialStartDate);
    setEndDate(project.endDate ? formatDate(project.endDate) : initialEndDate);
    setPriority(project.priority);
  }

  //Call this function after setStatusMessag to autoHide alert message
  function autoHideAlert() {
    setTimeout(() => {
      setStatusMessage({
        ...statusMessage,
        show: false
      });
    }, 5000);
  }

  function resetFormState() {
    setProjectName('');
    setStartDate(initialStartDate);
    setEndDate(initialEndDate);
    setPriority('');
    setDateRequired(false);
    setEditMode(false);
  }

  // function onCheckboxChange(event) {
  //   if (event.target.checked) {
  //     setDateRequired(true);
  //   } else {
  //     setDateRequired(false);
  //   }
  // }

  function getAfterDate(num) {
    return moment().add(num, 'day');
  }

  function formatDate(date) {
    return moment(date).format('YYYY-MM-DD');
  }

  useEffect(() => {
    fetchAllProjects();
  }, [projectUpdated]);

  const formik = useFormik({
    enableReinitialize: true,
    initialValues: {
      projectId: projectId,
      projectName: projectName,
      startDate: startDate,
      endDate: endDate,
      priority: priority,
      dateRequired: dateRequired
    },
    validationSchema: Yup.object({
      projectName: Yup.string().required('Please enter Project Name'),
      startDate: Yup.date().when('dateRequired', {
        is: true,
        then: Yup.date().required('Please enter start date'),
        otherwise: Yup.date()
      }),
      endDate: Yup.date().when('dateRequired', {
        is: true,
        then: Yup.date()
          .required('Please enter end date')
          .test('compare', 'End date should be greater than Start date', function(endDate) {
            return endDate > this.parent.startDate;
          }),
        otherwise: Yup.date()
      }),
      priority: Yup.number().required('Please enter Priority'),
      dateRequired: Yup.boolean()
    }),
    onSubmit: async value => {
      if (!value.dateRequired) {
        delete value.startDate;
        delete value.endDate;
      }
      try {
        const resp = !editMode ? await addNewProject(value) : await updateProjectById(value);
        setProjectUpdated(true);
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: resp.message,
          variant: 'success'
        });
      } catch (err) {
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: err,
          variant: 'danger'
        });
      }
      resetFormState();
      formik.resetForm();
    }
  });

  return (
    <>
      <Row>
        <Col className="mt-3">
          <Alert
            variant={statusMessage.variant}
            show={statusMessage.show}
            onClose={() => {
              setStatusMessage({ ...statusMessage, show: false });
            }}
            dismissible
          >
            {statusMessage.message}
          </Alert>
          <form onSubmit={formik.handleSubmit}>
            <Form.Group as={Row}>
              <Form.Label column sm="2">
                Project Name:
              </Form.Label>
              <Col sm="10">
                <FormControl
                  required
                  placeholder="ProjectName"
                  name="projectName"
                  errors={formik.errors.projectName}
                  className={
                    formik.touched.projectName
                      ? formik.errors.projectName
                        ? 'is-invalid mb-2 mt-2'
                        : 'is-valid mb-2 mt-2'
                      : 'mb-2 mt-2'
                  }
                  {...formik.getFieldProps('projectName')}
                ></FormControl>
                <FormControl.Feedback type="invalid">
                  {formik.errors.projectName}
                </FormControl.Feedback>
              </Col>
            </Form.Group>

            <Form.Check
              type="checkbox"
              name="dateRequired"
              value={formik.values.dateRequired}
              checked={formik.values.dateRequired}
              label="Set Start and End date"
              //onChange={formik.onChange}
              //onBlur={formik.onBlur}
              // errors={formik.errors.dateRequired}
              // className={
              //   formik.touched.dateRequired
              //     ? formik.errors.dateRequired
              //       ? 'is-invalid mb-2 mt-2'
              //       : 'is-valid mb-2 mt-2'
              //     : 'mb-2 mt-2'
              // }
              {...formik.getFieldProps('dateRequired')}
              // {...formik.getFieldMeta('dateRequired')}
            />
            {/* <FormControl.Feedback type="invalid">{formik.errors.dateRequired}</FormControl.Feedback> */}
            <FormControl
              placeholder="Start Date"
              name="startDate"
              type="date"
              disabled={!formik.values.dateRequired}
              required={false}
              errors={formik.errors.startDate}
              className={
                formik.touched.startDate
                  ? formik.errors.startDate
                    ? 'is-invalid mb-2 mt-2'
                    : 'is-valid mb-2 mt-2'
                  : 'mb-2 mt-2'
              }
              {...formik.getFieldProps('startDate')}
            ></FormControl>
            <FormControl.Feedback type="invalid">{formik.errors.startDate}</FormControl.Feedback>
            <FormControl
              placeholder="End Date"
              name="endDate"
              type="date"
              disabled={!formik.values.dateRequired}
              required={false}
              errors={formik.errors.endDate}
              className={
                formik.touched.endDate
                  ? formik.errors.endDate
                    ? 'is-invalid mb-2 mt-2'
                    : 'is-valid mb-2 mt-2'
                  : 'mb-2 mt-2'
              }
              {...formik.getFieldProps('endDate')}
            ></FormControl>
            <FormControl.Feedback type="invalid">{formik.errors.endDate}</FormControl.Feedback>
            <FormControl
              placeholder="priority"
              name="priority"
              required
              errors={formik.errors.priority}
              className={
                formik.touched.priority
                  ? formik.errors.priority
                    ? 'is-invalid mb-2 mt-2'
                    : 'is-valid mb-2 mt-2'
                  : 'mb-2 mt-2'
              }
              {...formik.getFieldProps('priority')}
            ></FormControl>
            <FormControl.Feedback type="invalid">{formik.errors.priority}</FormControl.Feedback>

            <div className="float-right">
              <Button
                variant="primary"
                disabled={!formik.isValid || !formik.dirty}
                type="submit"
                className="ml-2 mr-2"
              >
                {editMode ? 'Update' : 'Add'}
              </Button>
              <Button
                variant="outline-secondary"
                onClick={() => {
                  resetFormState();
                  formik.resetForm();
                }}
              >
                Reset
              </Button>
            </div>
          </form>
        </Col>
      </Row>
      <hr />
      <Row>
        <Col>
          {projects.length === 0 && (
            <Alert variant="warning">Projects are not available in database!!</Alert>
          )}
          <ListGroup>
            {projects.map(project => {
              return (
                <ListGroup.Item className="project-list" key={project._id}>
                  <div>
                    <p>Project : {project.projectName}</p>
                    <p>Start Date : {project.startDate && formatDate(project.startDate)}</p>
                    <p>End Date : {project.endDate && formatDate(project.endDate)}</p>
                    <p>Priority : {project.priority}</p>
                  </div>
                  <div>
                    <Button variant="outline-primary" onClick={() => editProject(project)}>
                      Update
                    </Button>
                    <br />
                    <Button
                      variant="outline-secondary"
                      className="mt-2"
                      onClick={() => {
                        deleteProject(project);
                      }}
                    >
                      Suspend
                    </Button>
                  </div>
                </ListGroup.Item>
              );
            })}
          </ListGroup>
        </Col>
      </Row>
    </>
  );
};
export default AddProject;
