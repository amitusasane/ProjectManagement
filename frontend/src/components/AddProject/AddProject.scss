.project-list.list-group-item {
  display: flex;
  align-items: center;
  justify-content: space-between;
}
.modal-label {
  border: 1px solid #ced4da;
  flex: 1 1 auto;
  align-self: stretch;
  padding: 8px;
  margin-bottom: 0;
}
