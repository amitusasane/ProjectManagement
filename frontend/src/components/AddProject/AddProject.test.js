import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render, cleanup, wait } from '@testing-library/react';
import AddProject from './AddProject';

describe('Add Project Component', () => {
  afterEach(cleanup);
  it('Should render AddProject component', async () => {
    const { container } = render(
      <Router>
        <AddProject />
      </Router>
    );
    await wait(() => {
      expect(container).toMatchSnapshot();
    });
  });

  // it('Should render Add Project component', () => {
  //   expect(true).toBe(true);
  // });
});
