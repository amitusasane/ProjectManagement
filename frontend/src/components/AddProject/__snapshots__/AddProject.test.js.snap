// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Project Component Should render AddProject component 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="mt-3 row"
    >
      <div
        class="col"
      >
        <h3>
          Manage Project
        </h3>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="mt-3 col"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              class="form-label"
            >
              Project Name
            </label>
            <input
              class="mb-2 mt-2 form-control"
              name="projectName"
              placeholder="ProjectName"
              required=""
              value=""
            />
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="form-check"
          >
            <input
              class="form-check-input"
              name="dateRequired"
              type="checkbox"
              value="false"
            />
            <label
              class="form-check-label"
              title=""
            >
              Set Start and End date
            </label>
          </div>
          <label
            class="mt-2 form-label"
          >
            Start Date
          </label>
          <input
            class="mb-2 mt-2 form-control"
            disabled=""
            name="startDate"
            placeholder="Start Date"
            type="date"
            value="2019-12-14"
          />
          <div
            class="invalid-feedback"
          />
          <label
            class="form-label"
          >
            End Date
          </label>
          <input
            class="mb-2 mt-2 form-control"
            disabled=""
            name="endDate"
            placeholder="End Date"
            type="date"
            value="2019-12-15"
          />
          <div
            class="invalid-feedback"
          />
          <label
            class="mt-2 form-label"
          >
            Priority
          </label>
          <input
            class="mb-2 mt-2 form-control"
            max="30"
            min="0"
            name="priority"
            placeholder="priority"
            step="1"
            type="range"
            value=""
          />
          <div
            class="invalid-feedback"
          />
          <label
            class="form-label"
          >
            Search Manager
          </label>
          <div
            class="input-group"
          >
            <label
              class="modal-label form-label"
            />
            <input
              class="mb-2 mt-2 form-control-plaintext"
              hidden=""
              name="manager"
              readonly=""
              required=""
              value=""
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-outline-dark"
                type="button"
              >
                Search Manager
              </button>
            </div>
          </div>
          <div
            class="invalid-feedback"
          />
          <div
            class="float-right mt-4"
          >
            <button
              class="ml-2 mr-2 btn btn-dark"
              data-testid="updateProject"
              disabled=""
              type="submit"
            >
              Add
            </button>
            <button
              class="btn btn-light"
              type="button"
            >
              Reset
            </button>
          </div>
        </form>
      </div>
    </div>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-sm-6 col-12"
      >
        <input
          class="mb-4 form-control"
          placeholder="Search"
        />
      </div>
      <div
        class="col-sm-6 col-12"
      >
        Sort By:
        <button
          class="ml-2 btn btn-outline-dark"
          data-testid="sortStartDate"
          type="button"
        >
          Start Date
        </button>
        <button
          class="ml-2 btn btn-outline-dark"
          data-testid="sortEndDate"
          type="button"
        >
          End Date
        </button>
        <button
          class="ml-2 btn btn-outline-dark"
          data-testid="sortPriority"
          type="button"
        >
          Priority
        </button>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="list-group"
        >
          <div
            class="project-list list-group-item"
          >
            <div>
              <p>
                <label>
                  Project : 
                </label>
                Project 01
              </p>
              <p>
                <label>
                  Start Date : 
                </label>
                2019-12-13
              </p>
              <p>
                <label>
                  End Date : 
                </label>
                2019-12-20
              </p>
              <div>
                <span
                  class="mr-2"
                >
                  <label>
                    No of tasks: 
                  </label>
                  1
                </span>
              </div>
              <div>
                <span
                  class="mr-2"
                >
                  <label>
                    Completed: 
                  </label>
                  1
                </span>
              </div>
            </div>
            <div
              class="list-group"
            >
              <div
                class="list-group-item"
              >
                Priority : 
                24
              </div>
            </div>
            <div>
              <button
                class="btn btn-dark"
                type="button"
              >
                Update
              </button>
              <br />
              <button
                class="mt-2 btn btn-light"
                type="button"
              >
                Suspend
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
